### DISCUSSION TITLE
The language server extension is defaulting to using "didOpen" events to request code files that were not actually opened.

### DISCUSSION THREAD

**Discussion by Dongyang0810** [microsoft/vscode-discussions#2426] [Extension Development QnA]
I am developing a static code analysis extension using the Language Server Protocol. When I start a VSCode window to test the static code analysis extension, even though no code files are opened in the VSCode debugging window, the LSP language server still receives a "didOpen" request containing the file URL of some code file that was opened during a previous debugging session.

How to solve this problem: I hope that when opening a VSCode window, if no file is actually open, the "didOpen" event for previously closed files in VSCode will not be sent.

**Comment by robconery** [comment]
I'm not an expert in this, but I _think_ this issue happens because VSCode restores previously opened files, even if they appear closed. The LSP server receives `didOpen` events for these files. To prevent this in _in general_:

In `settings.json`:

```json
   {
     "workbench.editor.restoreWhenOpening": "none"
   }
```

You can also check to see if the file was reopened from within your extension:

```
connection.onDidOpenTextDocument((params) => {
    if (isRestoredFile(params.textDocument.uri)) {
        return; // Ignore restored files
    }
    // Handle normal file opening logic
});
```

Finally - you can also check if a file was reopened by using `workspace.onDidOpenTextDocument` event and adding some logic there.
